dalm/training/generator_only/trainer.py

```diff
@@ -0,0 +1,274 @@
+import os
+from typing import Optional
+
+import torch
+from accelerate import Accelerator
+from datasets import load_dataset
+from peft import LoraConfig
+from tqdm import tqdm
+from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, TrainingArguments
+
+from trl import SFTTrainer
+from trl.trainer import ConstantLengthDataset
+
+import argparse
+
+
+def create_datasets(
+    dataset_name: str,
+    split: str,
+    size_valid_set: Optional[int],
+    streaming: bool,
+    shuffle_buffer: int,
+    seq_length: int,
+    num_workers: int,
+    tokenizer: AutoTokenizer,
+    formatting_func: callable,
+):
+    dataset = load_dataset(
+        dataset_name,
+        split=split,
+        num_proc=num_workers if not streaming else None,
+        streaming=streaming,
+    )
+    if streaming:
+        print("Loading the dataset in streaming mode")
+        valid_data = dataset.take(size_valid_set)
+        train_data = dataset.skip(size_valid_set)
+        train_data = train_data.shuffle(buffer_size=shuffle_buffer, seed=None)
+    else:
+        dataset = dataset.train_test_split(test_size=0.05, seed=None)
+        train_data = dataset["train"]
+        valid_data = dataset["test"]
+        print(f"Size of the train set: {len(train_data)}. Size of the validation set: {len(valid_data)}")
+
+    chars_per_token = chars_token_ratio(train_data, tokenizer, formatting_func)
+    print(f"The character to token ratio of the dataset is: {chars_per_token:.2f}")
+
+    train_dataset = ConstantLengthDataset(
+        tokenizer,
+        train_data,
+        formatting_func=formatting_func,
+        infinite=True,
+        seq_length=seq_length,
+        chars_per_token=chars_per_token,
+    )
+    valid_dataset = ConstantLengthDataset(
+        tokenizer,
+        valid_data,
+        formatting_func=formatting_func,
+        infinite=False,
+        seq_length=seq_length,
+        chars_per_token=chars_per_token,
+    )
+    return train_dataset, valid_dataset
+
+
+def chars_token_ratio(dataset, tokenizer, formatting_func, nb_examples=400):
```

Comment:
What does `nb` stand for?  If "number", then I think `num_examples` is a bit clearer